lib/autoBlock.js

```diff
@@ -0,0 +1,215 @@
+var _ = require('lodash')
+var async = require('async')
+var httpStatus = require('http-status')
+
+var utils = require('./utilities')
+
+// TODO reject invalid options/flags (e.g., errorDeafults typos)
+
+// Runs a block using common block/controller pattern
+function runBlock (config, done) {
+  // Set config.done if not provided
+  if (!config.done) {
+    config.done = done || (() => {
+    })
+  }
+
+  if (!config.data) {
+    config.data = {}
+  }
+
+  // Setup noop logging hooks
+  if (!config.onStart) {
+    config.onStart = () => {
+    }
+  }
+
+  if (!config.onSuccess) {
+    config.onSuccess = () => {
+    }
+  }
+
+  if (!config.onFailure) {
+    config.onFailure = () => {
+    }
+  }
+
+  if (!config.onStartStep) {
+    config.onStartStep = () => {
+    }
+  }
+
+  if (!config.onFinishStep) {
+    config.onFinishStep = () => {
+    }
+  }
+
+  config.onStart(config.data)
+
+  // Use optionsMapping and optionsDefaults to build out options if none were explicitly provided on the block
+  if (!config.block.options) {
+    var options = utils.mappingBlock(config.data, config.optionsMapping, config.optionsDefaults)
+    config.block['options'] = {
+      value: options
+    }
+  }
+
+  var autoBlock = buildBlock(config.block, config)
+  var autoCb = (error, results) => finish(error, results, config)
+  try {
+    async.auto(autoBlock, autoCb)
+  } catch (e) {
+    autoCb(e)
+  }
+}
+
+// Resolve the error and results
+function finish (error, results, config) {
+  var data = utils.resultBlock(config, error, results)
+
+  // Map response if desired; can access raw error object
+  if (config.responseMapping) {
+    data.response = utils.mappingBlock(data, config.responseMapping)
+  }
+
+  // Map error if desired; can access response object
+  if (error && !error.requestId && (config.errorMapping || config.errorDefaults)) {
+    error = utils.mapError(data, config.errorMapping, config.errorDefaults)
+  }
+
+  var status = _.get(error, 'data.status')
+  if (error && error.message && status) {
+    error.message = (httpStatus[status] || status) + `: ${error.message}`
+  }
+
+  if (error) {
+    config.onFailure(error, data)
+  } else {
+    config.onSuccess(data.response, data)
+  }
+
+  var done = config.done
+  if (done) {
+    // Lambda will automatically retry if you send the error back
+    if (!_.get(error, 'data.retry', true)) {
+      error = null
+    }
+    process.nextTick(() => done(error, data.response))
```

Comment:
I feel like this should have a counter or timeout or something, otherwise how do you prevent infinite looping from an error?